### DISCUSSION TITLE
mock_engine creates different DDL for reflected tables and those defined in code.

### DISCUSSION THREAD

**Discussion by volcan01010** [sqlalchemy/sqlalchemy#11737] [Usage Questions] [answered]
## Summary

I am trying to use the `mock_engine` and dialects in SQLAlchemy to create Oracle DDL from tables reflected from a SQLite database.  However, the resulting SQL is invalid and looks more like DDL for the source SQLite database.  This is not the case when I create the DDL from hand-crafted SQLAlchemy models.

## Description of the issue

The code below demonstrates my issue.  I am most interested in the Oracle dialect, but the code prints the DDL in various dialects to show how they change.

### DDL output from tables created from SQLAlchemy models

This script creates a simple SQLAlchemy table model, then uses a `mock_engine` to print the DDL in different dialects.

```python
# sqla_tables.py
from sqlalchemy import (
    MetaData,
    Table,
    Column,
    Integer,
    String,
    create_mock_engine,
)

from sqlalchemy.schema import Table, ExecutableDDLElement
from sqlalchemy.dialects.oracle.oracledb import OracleDialect_oracledb
from sqlalchemy.dialects.postgresql.psycopg2 import PGDialect_psycopg2
from sqlalchemy.dialects.sqlite.pysqlite import SQLiteDialect_pysqlite


# Define table as SQLAlchemy model and add to MetaData class
metadata = MetaData()

test_table = Table(
    'test', metadata,
    Column('id', Integer, primary_key=True),
    Column('text_field', String),
    )


# Print DDL in dialects


def print_dialects(sql: ExecutableDDLElement, *multiparams, **params):
    print("-- original:" + str(sql))
    print("-- oracle: " + str(sql.compile(dialect=OracleDialect_oracledb())))
    print("-- postgres: " + str(sql.compile(dialect=PGDialect_psycopg2())))
    print("-- sqlite: " + str(sql.compile(dialect=SQLiteDialect_pysqlite())))
    

mock_engine = create_mock_engine(url='oracle+oracledb://', executor=print_dialects)
metadata.create_all(bind=mock_engine)
```

The output is below.

```
$ python sqla_tables.py
-- original:
CREATE TABLE test (
        id INTEGER NOT NULL,
        text_field VARCHAR,
        PRIMARY KEY (id)
)


-- oracle:
CREATE TABLE test (
        id INTEGER NOT NULL,
        text_field VARCHAR2,
        PRIMARY KEY (id)
)


-- postgres:
CREATE TABLE test (
        id SERIAL NOT NULL,
        text_field VARCHAR,
        PRIMARY KEY (id)
)


-- sqlite:
CREATE TABLE test (
        id INTEGER NOT NULL,
        text_field VARCHAR,
        PRIMARY KEY (id)
)
```

The DDL is created with a type of `VARCHAR` for the 'text_field' in most dialects.  I can see some differences, however, as the Oracle version includes VARCHAR2 instead of VARCHAR and the PostgreSQL version has SERIAL for the 'id' column.  I would have expected the type to be `TEXT` for PostgreSQL and SQLite.

### DDL output from tables reflected from SQLite database

This script creates a SQLite database with a table with the same structure as above.  It then reflects it and uses a `mock_engine` to print the DDL in different dialects.

```python
# reflected_tables.py
from pathlib import Path
import sqlite3

from sqlalchemy import (
    MetaData,
    create_engine,
    create_mock_engine,
)

from sqlalchemy.schema import ExecutableDDLElement
from sqlalchemy.dialects.oracle.oracledb import OracleDialect_oracledb
from sqlalchemy.dialects.postgresql.psycopg2 import PGDialect_psycopg2
from sqlalchemy.dialects.sqlite.pysqlite import SQLiteDialect_pysqlite

# Create a new SQLite database with single table
Path('test.db').unlink(missing_ok=True)
with sqlite3.connect('test.db') as conn:
    cursor = conn.cursor()
    cursor.execute("""
        CREATE TABLE test (
            id INTEGER PRIMARY KEY,
            text_field TEXT);
        """)

# Reflect the table into MetaData class
engine = create_engine(url="sqlite:///test.db")
metadata = MetaData()
metadata.reflect(bind=engine)

# Print DDL in dialects


def print_dialects(sql: ExecutableDDLElement, *multiparams, **params):
    print("-- original:" + str(sql))
    print("-- oracle: " + str(sql.compile(dialect=OracleDialect_oracledb())))
    print("-- postgres: " + str(sql.compile(dialect=PGDialect_psycopg2())))
    print("-- sqlite: " + str(sql.compile(dialect=SQLiteDialect_pysqlite())))
    

mock_engine = create_mock_engine(url='oracle+oracledb://', executor=print_dialects)
metadata.create_all(bind=mock_engine)
```

The output is below.

```
$ python reflected_tables.py
-- original:
CREATE TABLE test (
        id INTEGER,
        text_field TEXT,
        PRIMARY KEY (id)
)


-- oracle:
CREATE TABLE test (
        id INTEGER,
        text_field TEXT,
        PRIMARY KEY (id)
)


-- postgres:
CREATE TABLE test (
        id SERIAL,
        text_field TEXT,
        PRIMARY KEY (id)
)


-- sqlite:
CREATE TABLE test (
        id INTEGER,
        text_field TEXT,
        PRIMARY KEY (id)
)
```

This time the 'text_field' is of type TEXT everywhere.  In my case, this means the the DDL for Oracle doesn't work, because TEXT is not a valid data type there.  In the previous example, the 'text_field' was set to VARCHAR2 for Oracle, but this hasn't happened this time.  The dialects are having some effect, however, because the PostgreSQL 'id' column has SERIAL type.

The NOT NULL constraint on the 'id' field is not explicitly defined this time, although it is implied by the PRIMARY KEY constraint.

## Discussion / Questions

I did not expect the output DDL to be different depending on the source of the table model.

I have been unable to find the reason for the difference.  Creating a fresh `MetaData` class and add the reflected tables to it doesn't change things.  I have looked through the attributes of the created `Table` and `Column` classes but I can't see anything that would "remember" that the tables were reflected SQLite tables.

It is interesting that the `compile` function only makes slight modifications to the DDL and that the output mainly depends on the original form.

I think that generated Oracle DDL with `TEXT` as the 'text_field' type is a **bug**, as this should never be the case because it is not a valid type in Oracle.

Is there a way that I can alter the `metadata` or `Table` attributes for the reflected tables so that they "forget" that they came from SQLite tables?

Or is there a way that I can make the `compile` more forceful, so that it changes the 'text_field' type whatever the original value was?  I would like to be able to go directly from reflected SQLite tables to valid Oracle DDL.

Thanks,
John

## Software versions

+ sqlachemy 2.0.30 (conda package: sqlalchemy=2.0.30=py312h5eee18b_0)
+ Python 3.12.3
+ Ubuntu 22.04.4 LTS

**Comment by CaselIT** [comment, marked as answer]
Hi,

This is expected. You are using text in the sqlite column and the reflected column would most likely be using the uppercase TEXT type.

All uppercase types do not get adapted to target ddl, so for oracle the correct thing in this case is to render TEXT.

you coul likely look into the column_reflect event to change the sqlite specific types that are reflected into more general ones, like string or text

**Reply by volcan01010** [reply]
Thank you.  I hadn't appreciated the [difference between CamelCase and UPPERCASE types](https://docs.sqlalchemy.org/en/20/core/type_basics.html#module-sqlalchemy.types).

I looked up the [column_reflect event](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DDLEvents.column_reflect) and used the instructions to write a function that replaces some UPPERCASE types with CamelCase types.

A working `reflected_tables_with_generic_types.py` script is below.  Hopefully it will be useful for anyone with a similar issue.

```python
# reflected_tables_with_generic_types.py
from pathlib import Path
import sqlite3

from sqlalchemy import (
    MetaData,
    create_engine,
    create_mock_engine,
    Integer,
    String,
    event,
)

from sqlalchemy.sql.sqltypes import INTEGER, TEXT
from sqlalchemy.schema import ExecutableDDLElement
from sqlalchemy.dialects.oracle.oracledb import OracleDialect_oracledb
from sqlalchemy.dialects.postgresql.psycopg2 import PGDialect_psycopg2
from sqlalchemy.dialects.sqlite.pysqlite import SQLiteDialect_pysqlite

# Create a new SQLite database with single table
Path('test.db').unlink(missing_ok=True)
with sqlite3.connect('test.db') as conn:
    cursor = conn.cursor()
    cursor.execute("""
        CREATE TABLE test (
            id INTEGER PRIMARY KEY,
            text_field TEXT);
        """)

# Create metadata class with customised column reflection
# to use more generic types.  See the following links for details:
# + https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DDLEvents.column_reflect
# + https://docs.sqlalchemy.org/en/20/core/type_basics.html#module-sqlalchemy.types
metadata = MetaData()

@event.listens_for(metadata, 'column_reflect')
def receive_column_reflect(inspector, table, column_info):
    # receives for all Table objects that are reflected
    # under this MetaData.
    generic_types = {
        INTEGER: Integer,
        TEXT: String
    }
    # Replace column type with generic one
    column_info['type'] = generic_types[type(column_info['type'])]


# Reflect the table into MetaData class
engine = create_engine(url="sqlite:///test.db")
metadata.reflect(bind=engine)

# Print DDL in dialects


def print_dialects(sql: ExecutableDDLElement, *multiparams, **params):
    print("-- original:" + str(sql))
    print("-- oracle: " + str(sql.compile(dialect=OracleDialect_oracledb())))
    print("-- postgres: " + str(sql.compile(dialect=PGDialect_psycopg2())))
    print("-- sqlite: " + str(sql.compile(dialect=SQLiteDialect_pysqlite())))
    

mock_engine = create_mock_engine(url='oracle+oracledb://', executor=print_dialects)
metadata.create_all(bind=mock_engine)
```

The output is

```
$ python reflected_tables.py
-- original:
CREATE TABLE test (
        id INTEGER,
        text_field VARCHAR,
        PRIMARY KEY (id)
)


-- oracle:
CREATE TABLE test (
        id INTEGER,
        text_field VARCHAR2,
        PRIMARY KEY (id)
)


-- postgres:
CREATE TABLE test (
        id SERIAL,
        text_field VARCHAR,
        PRIMARY KEY (id)
)


-- sqlite:
CREATE TABLE test (
        id INTEGER,
        text_field VARCHAR,
        PRIMARY KEY (id)
)
```
